sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java

```diff
@@ -398,6 +443,50 @@ public boolean isSupertypeOf(TypeName other) {
     }
   }
 
+  /**
+   * A LogicalType allows users to define a custom schema type.
+   *
+   * <p>A LogicalType is a way to define a new type that can be stored in a schema field using an
+   * underlying FieldType as storage. A LogicalType must specify a base FieldType used to store the
+   * data by overriding the {@link #getBaseType()} method. Usually the FieldType returned will be
+   * one of the standard ones implemented by Schema. It is legal to return another LogicalType, but
+   * the storage types must eventually resolve to one of the standard Schema types; it is not
+   * allowed to have LogicalTypes reference each other recursively via getBaseType. The {@link
+   * #toBaseType} and {@link #toInputType} should convert back and forth between the Java type for
+   * the LogicalType (InputT) and the Java type appropriate for the underlying base type (BaseT).
+   *
+   * <p>{@link #getIdentifier} must define a globally unique identifier for this LogicalType. A
```

Comment:
It seems that there is no verification on globally uniqueness of identifiers. What's the effect if identifiers are not unique (seems to me that there is no any effect)?

Reply:
Good question: I originally had a registration for LogicalTypes (all had to be registered and registration checks uniqueness). I removed it from this PR because I was trying to keep the PR smaller. If you think it's better I can add it back in, or I could keep it for a future PR.

Since identifier is the identity of the logical type, Beam code is allowed to check the identifier to see what the type is; if multiple types had the same identifier, this logic will break. In fact SQL does this in several places, and if a different LogicalType had the same identifier as the ones registered in CalciteUtils, things would break in strange ways.

Reply:
Thanks. I didn't notice IDENTIFIERs are used in Map as key. That explains uniqueness requirement.

Leaving it to future PR definitely is ok.